### PR TITLE
Add a secondary radius to the SpherePointer to disable far interaction earlier

### PR DESCRIPTION
**Pull request by lukastoenneMS**
## Overview

There is supposed to be a "dead zone" between disabling the far interaction and enabling (grabbable) near interaction with the SpherePointer. Previously there was a hardcoded margin in the IsNearObject function which has been removed, as well as a significantly larger base SphereCastRadius.

Introducing a configurable NearObjectMargin property adds back the dead zone, while keeping the SphereCastRadius at the desired smaller value (0.05m).

Also added a test to make sure the distance values work as expected.

## Changes
- Fixes: #5240 .

### CODE REVIEW

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointer.cs [hunk 1]
```diff
@@ -13,6 +13,26 @@ public class SpherePointer : BaseControllerPointer, IMixedRealityNearPointer
...
+        [Min(0.0f)]
+        [Tooltip("Additional distance between SphereCastRadius and NearObjectRadius")]
+        private float nearObjectMargin = 0.2f;
+        /// <summary>
+        /// Additional distance between <see cref="BaseControllerPointer.SphereCastRadius"/> and <see cref="NearObjectRadius"/>.
+        /// </summary>
+        /// <remarks>
+        /// This creates a dead zone in which far interaction is disabled before objects become grabbable.
+        /// </remarks>
+        public float NearObjectMargin => nearObjectMargin;
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointer.cs]: Does this (and `NearObjectRadius`) need to be public? They seem like values only used internally in this script.
Making things public for tests feels strange to me.

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/SpherePointerTests.cs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        // Keeping this low by default so the test runs fast. Increase it to be able to see hand movements in the editor.
+        private const int numFramesPerMove = 3;
+
+        private float colliderSurfaceZ;
+
+        // Initializes MRTK, instantiates the test content prefab and adds a pointer handler to the test collider
+        [SetUp]
+        public void SetUp()
+        {
+            TestUtilities.InitializeMixedRealityToolkit(true);
```
Comment (julenka) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/SpherePointerTests.cs]: Recommend using `PlayModeTestUtilities.Setup`, see [SlateTests](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/a5359f5ed4dacfd812b7bc983285693f96bc0c55/Assets/MixedRealityToolkit.Tests/PlayModeTests/SlateTests.cs#L31)

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/SpherePointerTests.cs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+using Microsoft.MixedReality.Toolkit.Utilities;
+using Microsoft.MixedReality.Toolkit.Input;
+
+namespace Microsoft.MixedReality.Toolkit.Tests
+{
+    // Tests to verify sphere pointer distances
+    public class SpherePointerTests
+    {
+        // Keeping this low by default so the test runs fast. Increase it to be able to see hand movements in the editor.
+        private const int numFramesPerMove = 3;
```
Comment (julenka) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/SpherePointerTests.cs]: We should do this for all tests, good idea.

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/SpherePointerTests.cs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            colliderSurfaceZ = centerZ - scale * 0.5f;
+
+            var collider = cube.GetComponentInChildren<Collider>();
+            Assert.IsNotNull(collider);
+        }
+
+        [TearDown]
+        public void TearDown()
+        {
+            TestUtilities.ShutdownMixedRealityToolkit();
```
Comment (julenka) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/SpherePointerTests.cs]: Recommend PlayModeTests.TearDown

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointer.cs (hunk 1 above)
Comment (julenka) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointer.cs]: I could thing of an example: what if you wanted to show an object *just out of reach*? So, at the point where you still have a hand ray but if you move just a bit closer the ray disappears. How would a consumer determine that distance without this var?

Comment (lukastoenneMS) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointer.cs]: I like to have at least a public getter for serializable fields. It seems wrong to be able to edit something in the inspector but having no way to at least _read_ that value at runtime.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointer.cs [hunk 5]
```diff
@@ -29,17 +49,31 @@ public bool IsNearObject
...
+        {
+            get
+            {
+                if (IsFocusLocked)
+                {
+                    return true;
+                }
+                else if (base.IsInteractionEnabled && TryGetNearGraspPoint(out Vector3 position))
+                {
+                    return UnityEngine.Physics.CheckSphere(position, SphereCastRadius, ~UnityEngine.Physics.IgnoreRaycastLayer);
```
[the remark this answers is not in the document]
Comment (lukastoenneMS) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointer.cs]: I'm just reproducing the previous behavior, which also did not include a check for specific components beyond the collider. I'd rather not make a potentially breaking change at this point.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointer.cs (hunk 1 above)
Comment (lukastoenneMS) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointer.cs]: It's an aspect of Unity i'm really not fond of: this fact that you have more access to settings in the editor by default than through the API. At least that's how our access policy with private fields plays out in practice.

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/SpherePointerTests.cs (hunk 3 above)
Comment (lukastoenneMS) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/SpherePointerTests.cs]: Yes, and preferably all tests should use a shared value (right now there are still a few tests using 30 or so steps). I'll make a separate PR for that.

Comment (lukastoenneMS) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/SpherePointerTests.cs]: #5358

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointer.cs (hunk 1 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointer.cs]: Works for me! Thanks for the additional context.

**Pull request merged**